README.md
<!-- badges: start -->
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN status](https://www.r-pkg.org/badges/version/sword)](https://CRAN.R-project.org/package=sword)
[![Lifecycle:
stable](https://lifecycle.r-lib.org/articles/figures/lifecycle-stable.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
<!-- badges: end -->

# Geomasking Tool 

The geomasking tool was developed by the New York State Department of Health (NYSDOH) Environmental Public Health Tracking (EPHT) Program with funding administered through Health Research Incorporated. The geomasker is maintained by Abigail Stamm at the NYSDOH. 

### Installing the geomasker

Run the code below in R to install GAT directly from GitHub.

``` r
# install devtools if you don't already have it
install.packages("devtools")
# install the development version of the geomasker from Github 
# with all required packages from CRAN
devtools::install_github("ajstamm/geomask", dependencies = TRUE, 
                         build_vignette = TRUE)
```

### Why create the geomasker


### How the geomasker works



### Disclaimer

The geomasker is provided as is. We welcome feedback on what worked well, suggestions for improvement, and bugs you encounter. Report all issues via the "Issues" tab or by emailing [NYSDOH EPHT](mailto:epht@health.ny.gov?subject=[Geomask in R]).

The geomasker was written in R-2.13.0 in Windows XP and was revised and converted to a package in R-4.2.0 in Windows 10 using RStudio-1.4.1103 and devtools-2.3.2. The latest version of GAT was compiled in R-4.2.2 and runs in R-4.0 through R-4.2.

If you are interested in the geomasker's history, including the original R script, visit the [archive](archive/). 





News.md
# geomask v1.4.0

* change dlg* to tcltk functions based on gatpkg
* clean up map and save it to PDF
* convert to R package


# geomask v1.3

* November, 2013 adapt to work with R 3.0.2
* changed guiDlgOpen to dlgOpen, guiDlgMessage to dlgMessage, dlgSave
* add $res after dlgOpen and dlgMessage where needed

# geomask v1.2

* July 20, 2011, add log file and adjust maximum distances if needed
* Aug 9, 2011 improve error catching for opening files

# geomask v1.1

* June 17, 2011, add minimum distance



testing/testing_files_setup.R
# create testing files for Onondaga County

ot <- tigris::tracts("NY", "Onondaga", year = 2010)

obg <- tigris::block_groups("NY", "Onondaga", year = 2010)

ow <- tigris::area_water("NY", "Onondaga", year = 2015)
om <- tigris::county_subdivisions("NY", "Onondaga", year = 2010)
ol <- tigris::landmarks("NY", "point", year = 2015)
t <- sf::st_contains(oc, ol) |> unlist()
  ol <- ol[t, ]
op <- tigris::landmarks("NY", "area", year = 2015)
  t <- sf::st_contains(oc, op) |> unlist()
  op <- op[t, ]

sf::write_sf(ot, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_tracts")
sf::write_sf(obg, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_blockgroups")
sf::write_sf(ow, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_water")
sf::write_sf(om, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_mcd")
sf::write_sf(ol, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_landmarkpoints")
sf::write_sf(op, dsn = "testing", driver = "ESRI shapefile",
             layer = "onondaga_landmarkpolygons")


# no native areas in Onondaga County
# n <- tigris::states(year = 2010) |>
#   dplyr::filter(STATEFP10 == 36)
# or <- tigris::native_areas(year = 2015)
# t <- sf::st_contains(n, or) |> unlist()
# or <- or[t, ]




R/geomask.R
#' The NYS Geomasking Tool (geomasker)
#'
#'
#' @description
#' A method for shifting point-level data to mask locations using user-defined
#' parameters with minimal knowledge of or experience using R.
#'
#' @name geomask
# @aliases geomasker
#'
#' @details
#'
#' Geomasking goals:
#' \itemize{
#'   \item simplify and standardize the process
#'   \item be in a format accessible to anyone
#'   \item not require prior experience with R
#'   \item be customizable for advanced R users
#' }
#'
# To run the default version of of the geomasker, follow the
# \href{../doc/tutorial.html}{Walkthrough} in the vignettes.
#
# For information about how the geomasker works, check out the
# \href{../doc/tech_notes.html}{technical notes}.
#'
#' The geomasker was made possible through funding made available to the New
#' York State Department of Health / HRI by the CDC through
#' grant # NUE1EH001341-03-01.
#' Its content is solely the responsibility of the authors and does not
#' necessarily represent the official views of the CDC.
#'
#' @author
#'  Abigail Stamm, Bureau of Environmental and Occupational
#'                 Epidemiology, New York State Department of Health \cr
#'  Gwen Babcock, Bureau of Environmental and Occupational
#'                Epidemiology, New York State Department of Health \cr
#'

# mostly, I copied dplyr at https://github.com/tidyverse/dplyr/blob/master/R/dplyr.r
# I borrowed authorship code from https://rdrr.io/cran/TLdating/src/R/TLdating-package.R
NULL


R/writeGMlog.R
#' Write the geomasker log
#'
#' This function writes a log of the geomasking process. It reports the
#' input and output datasets, variables and settings used, map projection,
#' and program start and end times.
#'
#' @param area         Spatial layer.
#' @param filevars     List of file names and paths. Of relevance to this
#'                     function are the filename, filein, and the combined save
#'                     path and save name, userout.
#' @param mysettings   List of system settings, including version, pkgdate,
#'                     starttime, and the booleans kml and exists.
#' @param maskvars     List of settings for calculating masked locations.
#' @param settingsfile R data file (*.Rdata) produced as part of the
#'                     geomasker's output. This file saves all settings for
#'                     the geomasker. Other options can be set to NULL only if
#'                     this option is defined.
#'
#' @details
#'
#' Notes on using the settingsfile option:
#'
#' 1. You will get an error if you moved the input shapefile before running
#'    the function with this option, since the function needs to access the
#'    input shapefile to recreate the log.
#' 2. Reading in an *.Rdata file from a previous version of the geomasker may
#'    result in incorrect elapsed time and geomasker version numbers being
#'    written to the log, or in errors that cause the log file to be
#'    incomplete, due to changes in settings saved to the *.Rdata file as the
#'    geomasker has evolved.
#'
#' @examples
#'
#' # if you run this example, it saves "geomasked_points.log" to your working
#' # directory
#'
#' if (interactive()) {
#'
#' maskvars <- list(
#'   min = 100,
#'   max = 1000,
#'   unit = "meters",
#'   projection = TRUE,
#'   point_id = "POINTID",
#'   bound_id = "GEOID10"
#' )
#'
#' filevars <- list(
#'   pointfile = "points",     # original locations file
#'   pointpath = getwd(),         # original locations path
#'   boundfile = "tracts",        # original boundary file
#'   boundpath = getwd(),         # original boundary path
#'   boundin = paste(getwd(), "tracts", sep = "/"),
#'   pathout = getwd(),           # save path
#'   fileout = "geomasked_points" # save file
#' )
#'
#' mysettings <- list(
#'   starttime = Sys.time(),
#'   version = "1.0",
#'   pkgdate = format(Sys.Date(), "%m-%d-%Y"),
#'   endtime = Sys.time() + 250,
#'   exists = TRUE,
#'   kml = FALSE
#' )
#'
#' ot <- tigris::tracts("NY", "Onondaga", year = 2010)
#'
#' writeGMlog(
#'   area = ot,
#'   filevars = filevars,
#'   mysettings = mysettings,
#'   maskvars = maskvars
#' )
#' }
#' @export


writeGMlog <- function(area = NULL, maskvars, filevars, mysettings = NULL,
                       settingsfile = NULL) {
  # rewrite to remove the need for the area file -
  # only need it for projection, which I could save in settings
  # set up ---
  if (!is.null(settingsfile)) {
    load(settingsfile)
    if (is.null(mysettings)) { # rerunning failed log
      mysettings <- list(version = utils::packageDescription("geomask")$Version,
                         pkgdate = utils::packageDescription("geomask")$Date,
                         starttime = Sys.time(),
                         endtime = Sys.time()) # needed for the log
    }
    mysettings$exists <- file.exists(paste(filevars$userout, "shp", sep = "."))
    area <- sf::st_read(dsn = filevars$pathin, layer = filevars$filein)
  }

  # fill in full list of names below; code will error otherwise
  vars <- names(data.frame(area))
  newvars <- c("point_id", "bound_id", "orig_lon", "orig_lat",
               "mask_lon", "mask_lat", "flag", "geometry")
  vars <- vars[!vars %in% newvars]
  oldvars <- ""
  for (i in 1:(length(vars)-3)) {
    oldvars <- paste0(oldvars, vars[i], ", ")
    if (i %% 6 == 0) {
      oldvars <- paste0(oldvars, "\n", paste(rep(" ", 22), collapse = ""))
    }
  }

  # begin log file ----
  logfile <- paste(filevars$fileout, "log", sep = ".")
  logfile <- paste(filevars$pathout, logfile, sep = "/")

  # settings ----
  logtext <- c("NYSDOH Geomasking Tool Log",
               "\n  Version & date:", mysettings$version, mysettings$pkgdate,
               "\n  Date run:", as.character(as.Date(Sys.time())),
               "\n  Time tool took to run:",
               round(difftime(mysettings$endtime, mysettings$starttime, units = "mins"),
                     digits = 2), "minutes")
  write(logtext, file = logfile, ncolumns = length(logtext), append = FALSE)

  # input file ----
  logtext <- c("\nInput file:          ", filevars$pointin,
               "\n  Projection:        ",
               sf::st_crs(area, parameters = TRUE)$proj4string,
               "\n  Field names:       ", oldvars,
               "\n  Identifier:        ", maskvars$point_id,
               "\n  Boundary file:     ", filevars$boundin,
               "\n  Boundary variable: ", maskvars$bound_id,
               "\nOutput file:         ", filevars$userout)
  write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)

  # masking settings ----
  logtext <- c("\nMasking settings:    ",
               "\n  Minimum distance:  ",
               format(maskvars$min, big.mark=",", scientific=FALSE),
               maskvars$unit,
               "\n  Minimum distance:  ",
               format(maskvars$max, big.mark=",", scientific=FALSE),
               maskvars$unit)
  write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)

  # files ----
  if (!mysettings$exists) {
    logtext <- "\n  The shapefiles failed to save. "
    write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)
  } else {
    # data dictionary ----
    logtext <- c("\n  Variables created by the geomasking process:",
                 paste0(filevars$fileout, ".shp"),
                 "\n    Variables created by GAT:",
                 "\n        point_id:", "duplicate of original locations identifier,",
                 maskvars$point_id,
                 "\n        bound_id:", "duplicate of boundaries identifier,",
                 maskvars$bound_id,
                 "\n        orig_lat:", "original location latitude",
                 "\n        orig_lon:", "original location longitude",
                 "\n        mask_lat:", "shifted location latitude",
                 "\n        mask_lon:", "shifted location longitude",
                 "\n        flag:    ", "number corresponds to the number of times buffers had ",
                 "\n                 ", "to be recalculated to create a valid area to select a ",
                 "\n                 ", "masking point using the equation",
                 "\n                 ", "  new maximum value = prior minimum value",
                 "\n                 ", "  new minimum value = half of prior minimum value",
                 "\n                 ", "e.g. flag = 2 means buffers had to be recalculated twice")
    write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)

    # saved files ----
    oshp <- paste(filevars$pointfile, "shp", sep = ".")
    bshp <- paste(filevars$boundfile, "shp", sep = ".")
    orshp <- paste(paste(filevars$pointfile, "old", sep = "_"), "shp", sep = ".")
    nshp <- paste(paste(filevars$pointfile, "new", sep = "_"), "shp", sep = ".")
    bshp <- paste(paste(filevars$pointfile, "buffer", sep = "_"), "shp", sep = ".")

    logtext <- c("All files have been saved to", filevars$pathout,
                 "\n  Original shapefile:  ", oshp,
                 "\n  Boundary shapefile:  ", bshp,
                 "\n  Original shapefile with both original and masked",
                 "latitude and longitude:",
                 "\n                       ", orshp,
                 "\n  Masked shapefile:    ", nshp,
                 "\n  Buffer shapefile:    ", bshp)
    write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)
  }
  logtext <- c("\n  Maps:                ",
               paste(filevars$fileout, "map.pdf", sep = "_"),
               "\n  Log file:            ",
               paste0(filevars$fileout, ".log"),
               "\n  R settings file:     ",
               paste(filevars$fileout, "settings.Rdata", sep = "_"))
  write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)
  if (mysettings$kml) {
    logtext <- c("\n  KML file (raw):    ",
                 paste0(filevars$fileout, ".kml"),
                 "\n  KMZ file (zipped): ",
                 paste0(filevars$fileout, ".kmz"))
  } else {
    logtext <- c(logtext, "\n  You chose not to write a KML file.")
  }

  write(logtext, file = logfile, ncolumns = length(logtext), append = TRUE)
  #end code to create log file
}



GMR-revised.R
#---- header ------------------------------------------------------------------
# revision author : Abigail Stamm
# original author : Gwen Babcock
# revision date : 4 January 2021
# last prior revision: V1.3 Nov 2013
#
#---- Gwen's notes ------------------------------------------------------------
# purpose : to jitter point coordinates
# 	input : point-level shapefile
#   process : randomly move the points a specified distance
#   output : shapefile, KML file, log, maybe map(s)
#
# platform :
# 	original : Programmed in R2.13.0 for Windows XP
#   latest prior : adapted to work in R3.0.2 for Windows 7
#   development time : June 2011 - November 2013
#   current: R4.2.3 on Windows 10
#
# libraries : sp, rgdal, tcltk, svDialogs (to be updated)
#
#---- future plans ------------------------------------------------------------
#   1. add features to keep the points within an area and to make the
#      distance the point is moved depend on population density
#   2. requires shapefiles to be in lat/long; look at function spTransform()
#      to change projection if needed
#   3. add code to check if input file is points or not; offer to calculate
#      centroids if it's polygons or lines?
#
#---- change log --------------------------------------------------------------
#   v1.1 June 17, 2011, add minimum distance
#   v1.2 July 20, 2011, add log file and adjust maximum distances if needed
#        Aug 9, 2011 improve error catching for opening files
#   v1.3 November, 2013 adapt to work with R 3.0.2
#	       changed guiDlgOpen to dlgOpen, guiDlgMessage to dlgMessage, dlgSave
#	       add $res after dlgOpen and dlgMessage where needed
#   v1.4 (in progress)
#        convert to package
#        change dlg* to tcltk functions based on gatpkg
#
#------------------------------------------------------------------------------


#---- libraries ----
devtools::load_all("R")
# confirmed required: tcltk, gatpkg, sf
# requires GAT because I don't feel like being redundant

#---- settings ----
# add ways to define and include point ID and boundary ID by name
mysettings <- list(version = "1.4.0", # packageDescription("geomask")$Version,
                   pkgdate = "2023-04-17", # packageDescription("geomask")$Date,
                   starttime = Sys.time()) # needed for the log

# pre-load lists
myshps <- list()
temp <- list(quit = FALSE, backopt = TRUE)
step <- 1
bgcol <- "thistle1"
buttoncol <- "plum2"
quitopt <- "Quit Geomasker"
settings <- NULL
settingsfile <- NULL

# for testing; read in settings file
# later, set up menu confirmation like in GAT
# settings <- paste("P:/Sections/EHS/Staff/ajs11/R/tools/geomaskTest/results",
#                   "save_test_settings.Rdata", sep = "/")


if (!is.null(settings)) {
  load(settings)
  step <- 6
  temp$flagconfirm <- TRUE
  filevars$userout <- paste0(filevars$userout, "_2")
  filevars$fileout <- paste0(filevars$fileout, "_2")
  myshps$point <- sf::st_read(dsn = filevars$pointpath,
                              layer = filevars$pointfile)
  myshps$bound <- sf::st_read(dsn = filevars$boundpath,
                              layer = filevars$boundfile)
} else {
  maskvars <- list(min = 100, max = 1000, unit = "meters")
  filevars <- list(pointin = "", boundin = "")
}





#---- progress bar ----
pb <- list(title = paste("NYSDOH Geomask Tool",
                         mysettings$version, mysettings$date),
           label = "NYSDOH Geomask Tool is running. Please wait for dialogs.")
tpb <- tcltk::tkProgressBar(title = pb$title, label = pb$label, min = 0,
                            max = 26, initial = 0, width = 400)

#---- user input ----

while(step < 7) { # gwen: get user input until finalized
  #---- step 1: request point shapefile ----
  # request or create point ID
  while (step == 1 & !temp$quit) {
    pb <- list(title = "NYSDOH Geomask Tool: identify shapefile",
               label = "Identifying and selecting the point-level shapefile.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title,
                            label = pb$label)

    temp$msg <- "Select the shapefile to mask"
    temppath <- gatpkg::locateGATshapefile(myfile = filevars$pointin,
                                           step = step, msg = temp$msg)
    filevars$pointin <- temppath$userin
    filevars$pointfile <- temppath$filein
    filevars$pointpath <- temppath$pathin
    rm(temppath)

    if (filevars$pointin == "cancel") {
      step <- 10
      temp$quit <- TRUE
    } else {
      # only care if file is point-level
      myshps$point <- sf::st_read(dsn = filevars$pointpath,
                                  layer = filevars$pointfile)

      # error checking
      if (!sum(sf::st_geometry_type(myshps$point) == "POINT") ==
           nrow(myshps$point)) {
        # message: wrong kind of shapefile; repeat dialog
        temp$msg <- paste("The shapefile", filevars$filein,
                          "contains non-point geographies.\n",
                          "Please select a new shapefile.")
        tcltk::tkmessageBox(title = "Shapefile invalid", type = "ok",
                            icon = "error", message = temp$msg)
      } else {
        # shift to error checking
        maskvars$projection <- sum(grepl("long", sf::st_crs(myshps$point))) > 0

        if (!maskvars$projection) {
          # if FALSE, convert to lat/long
          # WGS84 is common: EPSG code 4326
          # allow user to select projection from list?
          myshps$point <- sf::st_set_crs(myshps$point, 4326)
          myshps$point <- sf::st_transform(myshps$point, 4326)
        }

        temp$msg <- paste("Please select the variable that \nidentifies your points.")
        temp$hlp <- paste0("Select your point ID variable. \n",
                           "  \u2022  To continue,  click 'Next >'. \n",
                           "  \u2022  To quit the Geomasker, click 'Cancel'.")

        temp$items <- c()
        temp$data <- data.frame(myshps$point)
        temp$names <- names(temp$data)

        for (i in 1:(ncol(temp$data) - 1)) {
          t <- table(temp$data[, temp$names[i]])
          if (length(t) == nrow(temp$data)) {
            temp$items <- c(temp$items, temp$names[i])
          }
        }

        maskvars$point_id <-
          gatpkg::inputGATvariable(mylist = temp$items, instruction = temp$msg,
                                   title = "Point ID Variable", checkopt = "",
                                   checkbox = FALSE, help = temp$hlp, step = step,
                                   helppage = "inputGATvariable", myvar = NULL,
                                   check = "", backopt = temp$backopt,
                                   bgcol = bgcol, buttoncol = buttoncol,
                                   quitopt = quitopt)$myvar

        if (!temp$backopt) {
          step <- 7
        } else if (maskvars$point_id == "cancel") {
          step <- 10
          temp$quit <- TRUE
        } else {
          step <- step + 1
        }
      }
    }


  }

  #---- step 2: request boundary shapefile ----
  # request boundary variable/ID or create one
  while (step == 2 & !temp$quit) {
    pb <- list(title = "NYSDOH Geomask Tool: identify boundary file",
               label = "Identifying and selecting the boundary shapefile.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title,
                            label = pb$label)

    temp$msg <- "Select the boundary shapefile"
    temppath <- gatpkg::locateGATshapefile(myfile = filevars$boundin,
                                           step = step, msg = temp$msg)
    filevars$boundin <- temppath$userin
    filevars$boundfile <- temppath$filein
    filevars$boundpath <- temppath$pathin
    rm(temppath)

    if (filevars$boundin == "cancel") {
      step <- 10
      temp$quit <- TRUE
    } else {
      # only care if file is polygon-level
      myshps$bound <- sf::st_read(dsn = filevars$boundpath,
                                  layer = filevars$boundfile)
      # error checking
      temp$error <- FALSE
      if (!sum(sf::st_geometry_type(myshps$bound) == "POLYGON") ==
          nrow(myshps$bound)) {
        temp$error <- TRUE
      }
      if (temp$error) {
        temp$msg <- paste("The shapefile", filevars$boundin,
                          "contains non-polygon geographies. \n",
                          "Please select a new shapefile.")
        tcltk::tkmessageBox(title = "Shapefile invalid", type = "ok",
                            icon = "error", message = temp$msg)
        temp$error <- FALSE
      } else {
        myshps$bound <- sf::st_set_crs(myshps$bound, sf::st_crs(myshps$point))
        myshps$bound <- sf::st_transform(myshps$bound, sf::st_crs(myshps$point))

        # myshps$bound <- dplyr::mutate(myshps$bound, bound_id = dplyr::row_number())
        # myshps$bound <- dplyr::select(myshps$bound, bound_id)

        msg <- paste("Please select the variable that \nidentifies boundaries within",
                     "\nwhich your points should be relocated.")
        hlp <- paste0("Select your boundary variable. \n",
                      "  \u2022  To continue,  click 'Next >'. \n",
                      "  \u2022  To quit the Geomasker, click 'Cancel'.")

        temp$items <- c()
        temp$data <- data.frame(myshps$bound)
        temp$names <- names(temp$data)

        for (i in 1:(ncol(temp$data) - 1)) {
          t <- table(temp$data[, temp$names[i]])
          if (length(t) == nrow(temp$data)) {
            temp$items <- c(temp$items, temp$names[i])
          }
        }

        maskvars$bound_id <-
          gatpkg::inputGATvariable(mylist = temp$items, instruction = temp$msg,
                                   title = "Boundary Variable", checkopt = "",
                                   checkbox = FALSE, help = temp$hlp, step = step,
                                   helppage = "inputGATvariable", myvar = NULL,
                                   check = "", backopt = temp$backopt,
                                   bgcol = bgcol, buttoncol = buttoncol,
                                   quitopt = quitopt)$myvar


        if (!temp$backopt) {
          step <- 7
        } else if (maskvars$point_id == "cancel") {
          step <- 10
          temp$quit <- TRUE
        } else {
          step <- step + 1
        }
      }
    }


  }

  #---- step 3: min/max distance ----
  while (step == 3 & !temp$quit) {
    pb <- list(title = "NYSDOH Geomask Tool: select distances",
               label = "Selecting the minimum and maximum distances.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title,
                            label = pb$label)

    templist <- selectGMdistances(step = step, min = maskvars$min,
                                  max = maskvars$max, unit = maskvars$unit,
                                  backopt = temp$backopt,
                                  bgcol = bgcol, buttoncol = buttoncol,
                                  quitopt = quitopt)

    maskvars$min <- as.numeric(gsub(",", "", templist$min))
    maskvars$max <- as.numeric(gsub(",", "", templist$max))
    maskvars$unit <- templist$unit

    # error checking
    if (maskvars$unit == "cancel") {
      step <- 10
      temp$quit <- TRUE
    } else if (maskvars$unit == "back") {
      step <- step - 1
    } else {
      # check that minimum < maximum and minimum > 0
      temp$msg <- ""
      if (maskvars$min < 0) {
        temp$msg <- "Minimum distance must be at least 0. \n"
      }
      if (maskvars$max < maskvars$min) {
        temp$msg <- paste0(temp$msg, "Maximum distance must be greater than",
                           "minimum distance. \n")
      }
      if (maskvars$max == maskvars$min) {
        temp$checkmsg <- paste(" The minimum and maximum distances are the",
                               "same. \n This will result in points being",
                               "moved a fixed distance. \n Are you sure you",
                               "want to proceed?")
        temp$check <- gatpkg::inputGATmessage(title = "Distances confirmation",
                                              msg = temp$checkmsg, step = step,
                                              buttonopt = "Fix settings",
                                              backopt = FALSE)
        if (temp$check == "cancel") {
          temp$msg <- paste0(temp$msg, "Minimum and maximum distances are the",
                             "same. \n")
        }
      }
      if (is.na(maskvars$min)) {
        temp$msg <- paste0(temp$msg, "The minimum distance entered is not a",
                           "number. \n")
      }
      if (is.na(maskvars$max)) {
        temp$msg <- paste0(temp$msg, "The maximum distance entered is not a",
                           "number. \n")
      }
      if (temp$msg != "") {
        temp$msg <- paste0(temp$msg, "Please reselect your distances.")
        tcltk::tkmessageBox(title = "Distance selections invalid", type = "ok",
                            icon = "error", message = temp$msg)
      } else {
        if (!temp$backopt) {
          step <- 7
        } else {
          step <- step + 1
        }
      }
    }
    if (step == 4) {
      # convert max and min values to meters
      if (maskvars$unit == "kilometers") {
        maskvars$min <- maskvars$min * 1000
        maskvars$max <- maskvars$max * 1000
      } else if (maskvars$unit == "miles") {
        maskvars$min <- maskvars$min * 1609.344
        maskvars$max <- maskvars$max * 1609.344
      } else if (maskvars$unit == "feet") {
        maskvars$min <- maskvars$min / 3.2808399
        maskvars$max <- maskvars$max / 3.2808399
      }
    }
    rm(templist)
  }

  #---- step 4: kml ----
  while (step == 4 & !temp$quit) {
    pb <- list(title = "NYSDOH GAT: save KML?",
               label = "Identifying whether to save a KML file.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title, label = pb$label)

    mysettings$kml <- gatpkg::saveGATkml(step = step, backopt = TRUE,
                                         bgcol = bgcol, buttoncol = buttoncol,
                                         quitopt = quitopt)

    if (mysettings$kml %in% c("Yes", "No")) {
      if (mysettings$kml == "Yes") {
        mysettings$kml <- TRUE # save the kml
      } else {
        mysettings$kml <- FALSE # save the kml
      }
      if (!temp$backopt) {
        step <- 7
      } else {
        step <- step + 1
      }
    } else if (mysettings$kml == "cancel") {
      temp$quit <- TRUE
      step <- 10
    } else {
      step <- step - 1
    }
  }

  #---- step 5: save location ----
  while (step == 5 & !temp$quit) {
    # identify the save files' name and location
    pb <- list(title = "NYSDOH GAT: identify save file",
               label = "Identifying the name and location of your save file.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title,
                            label = pb$label)

    saves <- gatpkg::saveGATfiles()
    filevars$userout <- saves$userout
    filevars$fileout <- saves$fileout
    filevars$pathout <- saves$pathout
    rm(saves)

    if (filevars$fileout == "cancel") {
      temp$quit <- TRUE
      step <- 10
    } else {
      if (!temp$backopt) {
        step <- 7
      } else {
        step <- step + 1
      }
    }
  }

  #---- step 6: confirmation ----
  while (step == 6 & !temp$quit) {
    pb <- list(title = "NYSDOH Geomask Tool: confirm settings",
               label = "Confirming your geomasking settings.")
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title,
                            label = pb$label)

    temp$cancel <- confirmGMsettings(maskvars = maskvars, filevars = filevars,
                                     savekml = mysettings$kml, step = step,
                                     bgcol = bgcol, buttoncol = buttoncol,
                                     quitopt = quitopt)


    if (temp$cancel %in% c("Yes", "None")) {
      step <- 7 # done with user input
      # add population file
    } else if (temp$cancel == "back") { # now irrelevant
      step <- step - 1 # go back one
    } else if (temp$cancel == "cancel") {
      step <- 7
      temp$quit <- TRUE
    } else if (grepl("[0-9]", temp$cancel)) {
      temp$backopt <- FALSE
      step <- as.numeric(gsub("[^0-9]", "", temp$cancel))
    }

  }
  mysettings$quit <- temp$quit
} # end while step
# rm(temp)

#---- automatic processing ----
# at this point, step = 7
if (!mysettings$quit) {
  # calculate points ----
  # Abby: need to add progress bar steps
  #       do this later after full program rewritten

  set.seed(mysettings$starttime)

  myshps <- calculateGMpoints(myshps = myshps, maskvars = maskvars)



  # plot(sf::st_geometry(myshps$point), col = "blue", pch = 20)
  # plot(sf::st_geometry(myshps$shifted), col = "red", add = TRUE, pch = 20)

  # created flag variable to bypass warnings here

  #---- step ?: plot points ----
  # add progress bar

  myplot <- plotGMcompare(bound = myshps$bound, original = myshps$point,
                          shifted = myshps$shifted, maskvars = maskvars)


  # save files ----
    # save shapefile(s) ----
  # write resulting shapefile(s)
  temp$oldfile = paste(filevars$fileout, "old", sep = "_")
  sf::write_sf(myshps$old_full, dsn = filevars$pathout,
               layer = temp$oldfile, driver = "ESRI shapefile")
  sf::write_sf(myshps$new_full, dsn = filevars$pathout,
               layer = filevars$fileout, driver = "ESRI shapefile")

    # save kml ----
  # write kml if desired
  if (mysettings$kml) { # now includes descriptions
    step <- step + 1
    pb$label = "Writing the KML file."
    tcltk::setTkProgressBar(tpb, value = step, title = pb$title, label = pb$label)

    gatpkg::writeGATkml(myshp = myshps$newpoint, filename = filevars$fileout,
                        filepath = filevars$pathout, myidvar = point_id)
  }

    # save plot(s) ----
  grDevices::pdf(paste0(filevars$userout, "plots.pdf"), onefile=TRUE,
                 width = 10, height = 7)
  grDevices::replayPlot(myplot)
  grDevices::dev.off() # need to close pdf file



    # save settingsfile ----
  save(file = paste0(filevars$userout, "settings.Rdata"),
       list = c("filevars", "maskvars", "mysettings"))


  # Abby: program rewritten to this point
  #       rewrite and/or reorder everything below this point

    # save log ----

  mysettings$endtime <- Sys.time()

  writeGMlog(area = myshps$point, maskvars = maskvars, filevars = filevars,
             mysettings = mysettings, settingsfile = settingsfile)







# write log (incomplete) ----
# shift log to function

################################################################################
#create a log file
#should contain input files, output file, min and max distances or sources (and units), date, run time
#(8 items)
################################################################################
#begin log file
logfile<-paste(filevars$userout, "txt", sep=".")

setStatusBar(paste("NYSDOH Geomasking Tool: Writing log file ",logfile))

logtext <- "NYSDOH Geomasking Tool log"
write("", file = logfile, ncolumns = length(logtext), append = TRUE)
#logtext<-c("The current date is ", format(Sys.Date()))
#write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)
logtext<-c("The current date and time are ", format(Sys.time()))
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

logtext<-c("The input point file is: ",userfile)
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)
logtext<-c("Number of points moved: ",length(mydata@data[,1]))
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

logtext<-c("The input boundary file is: ",userfileb)
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)
write("",file=logfile,ncolumns=length(logtext),append=TRUE)

if(userdistc[2]=="NONE"){
logtext=c("Minimum distance requested was ",userdistc[1])
}else{logtext=c("Minimum distance was obtained from field ",userdistc[2],userdistc[5])}
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

if(userdistc[4]=="NONE"){#be sure to include units=userdistc[5]
logtext=c("Maximum distance requested was ",userdistc[3],userdistc[5])
}else{
logtext=c("Maximum distance was obtained from field ",userdistc[4])}
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

logtext<-c("The time this program took to move the points: ",endtime-starttime)
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

logtext<-c("The output file(s): ",userout)
write(logtext,file=logfile,ncolumns=length(logtext),append=TRUE)

#end code to create log file
##############################################################################

setStatusBar(paste("NYSDOH Geomask Tool: find your results in: ",userpathout))



#---- save user settings ----
# for testing; move to end later
# save relevant objects
save(file = paste0(filevars$userout, "_settings.Rdata"),
     list = c("filevars", "maskvars", "mysettings"))






} else {
  # "else" occurs only if geomasking is cancelled in the input phase
  msg <- "You have chosen to cancel the Geomasker."
  tcltk::tkmessageBox(title = "Process cancelled",
                      message = msg, type = "ok", icon = "warning")
}




